regparser/builder.py

```diff
@@ -216,3 +226,81 @@ class EmptyCache(object):
     node, so it should not be cached."""
     def fetch_or_process(self, layer, node):
         return layer.process(node)
+
+
+def _serialize_xml_fields(node):
+    if node.source_xml is not None:
+        node.source_xml = etree.tostring(node.source_xml)
+
+
+def _deserialize_xml_fields(node):
+    if node.source_xml:
+        node.source_xml = etree.fromstring(node.source_xml)
+
+
+class Checkpointer(object):
+    """Save checkpoints during the build pipeline. Generally, a caller will
+    specify, a unique tag (a string) and a fallback function (for how to
+    compute it when there is no checkpoint). Calling checkpoint increment the
+    counter field, which is prefixed to the filename to limit the risk of
+    re-ordering collisions."""
+    def __init__(self, path):
+        self.counter = 0
+        self.path = path
+        self.suffix = ""
+        self.ignore_checkpoints = False
+        if not os.path.isdir(path):
+            os.makedirs(path)
+
+    def _filename(self, tag):
```

Comment:
I see some higher level tests below, I'd encourage you to write more tests around some of these lower-level functions.